src/physics/utils/pio_reader.F90

```diff
@@ -0,0 +1,751 @@
+    
+    implicit none
+    private
+  
+    public :: pio_reader_t
+
+    !Error code parameters
+    integer, parameter :: pio_inq_dim_id_err      = 1
+    integer, parameter :: pio_inq_dim_len_err     = 2
+    integer, parameter :: pio_inq_var_id_err      = 3
+    integer, parameter :: pio_inq_var_info_err    = 4
+    integer, parameter :: bad_var_rank_err        = 5
+    integer, parameter :: too_high_rank_err       = 6
+    integer, parameter :: pio_get_var_err         = 7
+    integer, parameter :: not_char_type_err       = 8
+
+    type :: sima_pio_fh_type
+        logical            :: is_file_open = .false.  !Is NetCDF file currently open?
+        type(file_desc_t)  :: pio_fh                  !PIO File handle type
+        character(len=cl)  :: file_path              !Local path to NetCDF file
+    end type
+
+    type, extends(ccpp_io_reader_t) :: pio_reader_t
+    private
+        type(sima_pio_fh_type) :: sima_pio_fh !PIO File handle type
+    contains
+        procedure :: open_netcdf_file => open_netcdf_file
+        procedure :: close_netcdf_file => close_netcdf_file
+        procedure :: get_netcdf_var_int => get_netcdf_var_int
+        procedure :: get_netcdf_var_real => get_netcdf_var_real
+        procedure :: get_netcdf_var_char => get_netcdf_var_char
+    end type pio_reader_t
+    
+contains
+
+    subroutine open_netcdf_file(this, file_path, errcode, errmsg)
+        use ioFileMod,        only: cam_get_file
+        use cam_pio_utils,    only: cam_pio_openfile
+        use pio,              only: PIO_NOWRITE
+
+        class(pio_reader_t), intent(inout)  :: this
+        character(len=*),    intent(in)  :: file_path
+        integer,             intent(out) :: errcode
+        character(len=*),    intent(out) :: errmsg
+        
+        character(len=cl)  :: local_file_path  !NetCDF file path on local file system
+
+        if(this%sima_pio_fh%is_file_open) then
+          errcode = 1
+          errmsg = "Trying to reuse pio_reader already used for: '"//this%sima_pio_fh%file_path
+        end if
+
+        if(file_path .eq. 'UNSET_PATH') then
+         errcode = 1
+         errmsg = "Found UNSET_PATH trying to open file"
+        end if
+  
+        call cam_get_file(file_path, local_file_path)
+        call cam_pio_openfile(this%sima_pio_fh%pio_fh, local_file_path, PIO_NOWRITE)
+  
+        !Set file handle metadata
+        this%sima_pio_fh%file_path    = local_file_path
+        this%sima_pio_fh%is_file_open = .true.
+  
+        !File was successfully opened
+        errcode = 0
+        errmsg = ''
+    end subroutine open_netcdf_file
+
+    subroutine close_netcdf_file(this, errcode, errmsg)
+        use pio, only: pio_closefile
+
+        class(pio_reader_t), intent(inout)  :: this
+        integer,             intent(out) :: errcode
+        character(len=*),    intent(out) :: errmsg
+
+        !Close NetCDF File:
+        call pio_closefile(this%sima_pio_fh%pio_fh)
```

Comment:
Should you first check if `is_file_open == .true.`?

Reply:
Good call!  Done!